deepspeed/runtime/zero/stage_1_and_2.py

```diff
@@ -833,6 +850,64 @@ def increment_value(dictionary, key):
     def overlapping_partition_gradients_reduce_epilogue(self):
         self.independent_gradient_partition_epilogue()
 
+    # Set the accum grad pointer to the correct tensor
+    def set_grad_accum(self, param):
+        # Simply point to the native pytorch grad
+        if not self.use_separate_grad_accum:
+            param.grad_accum = param.grad
+        # Accum the gradient in a separate tensor
+        else:
+            if param.grad is not None:
+                if param.grad_accum is None:
+                    param.grad_accum = param.grad.to(self.gradient_accumulation_dtype)
+                else:
+                    param.grad_accum.add_(
+                        param.grad.to(self.gradient_accumulation_dtype).view(
+                            param.grad_accum.shape))
+                # Don't clear if you still need the native grad for reduction (stage 2)
+                if self.use_grad_accum_for_reduction:
+                    param.grad = None
+
+    # Clear the tensor the reduction gradient attribute is pointing to
+    def clear_grad_reduc(self, param):
+        param.grad_reduc = None
+        if self.use_grad_accum_for_reduction:
+            param.grad_accum = None
+        else:
+            param.grad = None
+
+    # Point to the gradient tensor to use in reduction
+    def point_grad_reduc(self, param):
+        if self.use_grad_accum_for_reduction:
+            param.grad_reduc = param.grad_accum
+        else:
+            param.grad_reduc = param.grad
+
+    # Called after a memory defrag moves tensor object via copy
+    def point_grad_to_reduc(self, param):
+        if self.use_grad_accum_for_reduction:
+            param.grad_accum = param.grad_reduc
+        else:
+            param.grad = param.grad_reduc
+
+    def create_set_grad_accum_hooks(self):
```

Comment:
Hooks can incur runtime overheads that could be non-trivial for zero stage 1, so I would like to avoid them. Couldn't we call `set_grad_accum` for all params in the optimizer constructor?

Reply:
It could be replaced by nested for loops in the backward pass.

Reply:
Let's chat about this.